concurrency/RecoveryState.java

```diff
@@ -0,0 +1,961 @@
+         */
+        public synchronized int totalRecoverFiles() {
+            int total = 0;
+            for (File file : fileDetails.values()) {
+                if (file.reused() == false) {
+                    total++;
+                }
+            }
+            return total;
+        }
+
+
+        /**
+         * number of file that were recovered (excluding on ongoing files)
+         */
+        public synchronized int recoveredFileCount() {
+            int count = 0;
+            for (File file : fileDetails.values()) {
+                if (file.fullyRecovered()) {
+                    count++;
+                }
+            }
+            return count;
+        }
+
+        /**
+         * percent of recovered (i.e., not reused) files out of the total files to be recovered
+         */
+        public synchronized float recoveredFilesPercent() {
+            int total = 0;
+            int recovered = 0;
+            for (File file : fileDetails.values()) {
+                if (file.reused() == false) {
+                    total++;
+                    if (file.fullyRecovered()) {
+                        recovered++;
+                    }
+                }
+            }
+            if (total == 0 && fileDetails.size() == 0) {      // indicates we are still in init phase
+                return 0.0f;
+            }
+            if (total == recovered) {
+                return 100.0f;
+            } else {
+                float result = 100.0f * (recovered / (float) total);
+                return result;
+            }
+        }
+
+        /**
+         * total number of bytes in th shard
+         */
+        public synchronized long totalBytes() {
+            long total = 0;
+            for (File file : fileDetails.values()) {
+                total += file.length();
+            }
+            return total;
+        }
+
+        /**
+         * total number of bytes recovered so far, including both existing and reused
+         */
+        public synchronized long recoveredBytes() {
+            long recovered = 0;
+            for (File file : fileDetails.values()) {
+                recovered += file.recovered();
+            }
+            return recovered;
+        }
+
+        /**
+         * total bytes of files to be recovered (potentially not yet done)
+         */
+        public synchronized long totalRecoverBytes() {
+            long total = 0;
+            for (File file : fileDetails.values()) {
+                if (file.reused() == false) {
+                    total += file.length();
+                }
+            }
+            return total;
+        }
+
+        public synchronized long totalReuseBytes() {
+            long total = 0;
+            for (File file : fileDetails.values()) {
+                if (file.reused()) {
+                    total += file.length();
+                }
+            }
+            return total;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 859:865, 902:908.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.